fastly/origin_inspector.go

```diff
@@ -0,0 +1,146 @@
+	Latency10to50ms       uint64 `mapstructure:"latency_10_to_50ms"`
+	Latency50to100ms      uint64 `mapstructure:"latency_50_to_100ms"`
+	Latency100to250ms     uint64 `mapstructure:"latency_100_to_250ms"`
+	Latency250to500ms     uint64 `mapstructure:"latency_250_to_500ms"`
+	Latency500to1000ms    uint64 `mapstructure:"latency_500_to_1000ms"`
+	Latency1000to5000ms   uint64 `mapstructure:"latency_1000_to_5000ms"`
+	Latency5000to10000ms  uint64 `mapstructure:"latency_5000_to_10000ms"`
+	Latency10000to60000ms uint64 `mapstructure:"latency_10000_to_60000ms"`
+	Latency60000ms        uint64 `mapstructure:"latency_60000ms"`
+}
+
+// OriginMeta is the meta section returned for /metrics/origins responses
+type OriginMeta struct {
+	Start      string            `mapstructure:"start"`
+	End        string            `mapstructure:"end"`
+	Downsample string            `mapstructure:"downsample"`
+	Metric     string            `mapstructure:"metric"`
+	Limit      int               `mapstructure:"limit"`
+	NextCursor string            `mapstructure:"next_cursor"`
+	Sort       string            `mapstructure:"sort"`
+	GroupBy    string            `mapstructure:"group_by"`
+	Filters    map[string]string `mapstructure:"filters"`
+}
+
+// GetOriginMetricsInput is the input to an OriginMetrics request.
+type GetOriginMetricsInput struct {
+	ServiceID   string
+	Start       time.Time
+	End         time.Time
+	Metrics     []string
+	GroupBy     []string
+	Downsample  string
+	Hosts       []string
+	Datacenters []string
+	Regions     []string
+	Cursor      string
+}
+
+// GetOriginMetricsForService returns stats data based on GetOriginMetricsInput
+func (c *Client) GetOriginMetricsForService(i *GetOriginMetricsInput) (*OriginInspector, error) {
+	var resp interface{}
+	if err := c.GetOriginMetricsForServiceJSON(i, &resp); err != nil {
+		return nil, err
+	}
+
+	var or *OriginInspector
+	if err := decodeMap(resp, &or); err != nil {
+		return nil, err
+	}
+	return or, nil
+}
+
+// GetOriginMetricsForServiceJSON fetches Origin Inspector metrics for a single service and decodes the response
+// directly to the JSON struct dst.
+func (c *Client) GetOriginMetricsForServiceJSON(i *GetOriginMetricsInput, dst interface{}) error {
+	if i.ServiceID == "" {
+		return ErrMissingServiceID
+	}
+
+	p := "/metrics/origins/services/" + i.ServiceID
+
+	start := ""
+	if !i.Start.IsZero() {
+		start = strconv.FormatInt(i.Start.Unix(), 10)
+	}
+	end := ""
+	if !i.End.IsZero() {
+		end = strconv.FormatInt(i.End.Unix(), 10)
+	}
+
+	r, err := c.Get(p, &RequestOptions{
+		Params: map[string]string{
+			"start":      start,
+			"end":        end,
+			"downsample": i.Downsample,
+			"metric":     strings.Join(i.Metrics, ","),
+			"host":       strings.Join(i.Hosts, ","),
+			"datacenter": strings.Join(i.Datacenters, ","),
+			"region":     strings.Join(i.Regions, ","),
+			"cursor":     i.Cursor,
+		},
+	})
+	if err != nil {
+		return err
+	}
```

Comment:
I wonder if, because it's an LA (Limited Availability) feature, we should (or are able) to identify if the response is a 4xx (or 5xx) because the user doesn't have the feature enabled on their account? 

This would allow us to provide a clearer error response/remediation for the user, otherwise they may not understand why the API call failed if they didn't notice the LA status of the feature in the documentation.

I would imagine the worst case is there would be stale code that we'd need to remove once the feature goes GA (General Availability) and so we'd want to add a `TODO: ...` code comment for future maintainers.

Thoughts?

Reply:
For posterity: we discussed my comment internally and it seems there currently isn't a mechanism for detecting this _yet_ and when such a thing becomes possible we can update the logic to reflect that to give a better UX (user experience).

Reply:
Exactly. And also the `details` field from the API response will reflect why the error occurred, and this is propagated as part of the `error`. So I think, even longer term, we shouldn't need special handling for this case in the module, as users should be able to figure out fairly quickly why it's not working from the API response.